### PR TITLE
fix: nested swagger ref

### PR DESCRIPTION
**Pull request by peter-mouland**
related to https://github.com/fastify/fastify-swagger/pull/478 and https://github.com/fastify/fastify-swagger/pull/472

Fixes nested + complex ref's when using the swagger spec

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by peter-mouland** [comment]
hey @climba03003  @Eomm 
I've rebased from master to remove the conflicts, is there anything else we need to do to get the PR in?

File: lib/spec/swagger/index.js [hunk 1]
```diff
@@ -19,16 +19,19 @@ module.exports = function (opts, cache, routes, Ref, done) {
...
       ...(ref.definitions().definitions)
-    }
+    }, ref)
 
     // Swagger doesn't accept $id on /definitions schemas.
     // The $ids are needed by Ref() to check the URI so we need
     // to remove them at the end of the process
     Object.values(swaggerObject.definitions)
-      .forEach(_ => { delete _.$id })
+      .forEach(_ => {
```
Comment (climba03003) [inline lib/spec/swagger/index.js:31]: Can you move this logic inside `prepareSwaggerDefinitions`? I think it is not necessary to have two loop here.

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: lib/spec/swagger/index.js (hunk 1 above)
Comment (peter-mouland) [reply lib/spec/swagger/index.js]: done - and i moved the openAPI version to the same equivalent place

Comment (climba03003) [reply lib/spec/swagger/index.js]: Thanks

File: lib/spec/swagger/utils.js [hunk 2]
```diff
@@ -266,9 +266,29 @@ function prepareSwaggerMethod (schema, ref, swaggerObject) {
...
+      const _ = { ...definition }
+      const resolved = ref.resolve(_, { externalSchemas: [definitions] })
+
+      // Swagger doesn't accept $id on /definitions schemas.
+      // The $ids are needed by Ref() to check the URI so we need
+      // to remove them at the end of the process
+      delete resolved.$id
+      delete resolved.definitions
+
+      return {
```
Comment (climba03003) [inline lib/spec/swagger/utils.js:281]: I do not catch it in first place. But we should prevent a `reduce` loop and `spread` operator in here.
It create a new object each round of the loop.

```js
res[name] = resolved
return res
```

The above code should be enough.

Comment (climba03003) [reply lib/spec/swagger/utils.js:281]: Benchmark

```js
const { performance } = require('perf_hooks');
const arr = Array(10000).fill(0)

function speedSpread() {
  const start = performance.now()
  const obj = arr.reduce(function(obj) {
    return { ...obj, [Math.random()]: Math.random() } 
  }, {})
  return performance.now() - start
}

function speedPropForEach() {
  const start = performance.now()
  const obj = {}
  arr.forEach(function() {
    obj[Math.random()] = Math.random()
  })
  return performance.now() - start
}

function speedPropReduce() {
  const start = performance.now()
  const obj = arr.reduce(function(obj) {
    obj[Math.random()] = Math.random()
    return obj
  }, {})
  return performance.now() - start
}

const a = speedSpread()
const b = speedPropForEach()
const c = speedPropReduce()

console.log('speedSpread', a)
console.log('speedPropForEach', b)
console.log('speedPropReduce', c)
```

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: lib/spec/swagger/utils.js (hunk 2 above)
Comment (peter-mouland) [reply lib/spec/swagger/utils.js]: nice - updated for openAPI and swagger 👍

**Review by climba03003 — APPROVED** [review]
LGTM.

**Pull request merged**